### PR TITLE
chore: upgrade kubernetes dependencies

### PR DESCRIPTION
**Pull request by sagikazarmark**
Signed-off-by: Mark Sagi-Kazar <mark.sagikazar@gmail.com>

**What this PR does / why we need it**:

Upgrades the Kubernetes libraries to 1.19.2
Keeping libraries up to date is generally a good idea. I'm not aware of any major changes that might break backwards compatibility.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

The logr package needed a manual update because of container-runtime. Once 0.7.0 is out, we don't need to manually pin the logr library version.

Some other libraries have also been updated because they are also transitive dependencies.

### CODE REVIEW

**Comment by sagikazarmark** [comment]
Hm, can't see any obvious reason why the tests are breaking...

File: scripts/pre-commit.sh [hunk 1]
```diff
@@ -149,7 +149,7 @@ echo "Downloading e2e test dependencies"
 ./scripts/download-e2e-binaries.sh
 
 CREATE_INSECURE_REGISTRY=y CONFIGURE_INSECURE_REGISTRY_HOST=y \
-    KIND_TAG="v1.18.4" ./scripts/create-clusters.sh
+    KIND_TAG="v1.19.2" ./scripts/create-clusters.sh
```
Comment (jimmidyson) [inline scripts/pre-commit.sh:152]: There is only a [v1.19.1 tag available](https://hub.docker.com/r/kindest/node/tags?page=1&name=v1.19) here so please use that.
```suggestion
    KIND_TAG="v1.19.1" ./scripts/create-clusters.sh
```

File: go.mod [hunk 2]
```diff
@@ -3,26 +3,28 @@ module sigs.k8s.io/kubefed
...
+	github.com/spf13/cobra v1.0.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.6.1
-	k8s.io/api v0.18.6
-	k8s.io/apiextensions-apiserver v0.18.6
-	k8s.io/apimachinery v0.18.6
-	k8s.io/apiserver v0.18.6
-	k8s.io/client-go v0.18.6
-	k8s.io/component-base v0.18.6
+	k8s.io/api v0.19.2
```
Comment (jimmidyson) [inline go.mod:19]: Now that `v1.19.3` is out we could update to `v0.19.3` here and other dependencies. Would you mind taking care of that?

**Review by jimmidyson — CHANGES_REQUESTED** [review]
Thank you for this. Please fix the kind image and, if you don't mind, update to v0.19.3 dependencies, as well as running `go mod tidy` please to try to keep `go.sum` in a reasonable state.

File: scripts/pre-commit.sh (hunk 1 above)
Comment (sagikazarmark) [reply scripts/pre-commit.sh]: Done

File: go.mod (hunk 2 above)
Comment (sagikazarmark) [reply go.mod:19]: 1.19.3 is not available in all components yet (eg. in k8s.io/api). I guess we can hold off merging until it's available.

**Comment by sagikazarmark** [comment]
@jimmidyson thanks for the review. I addressed/replied to your comments.

I actually ran `go mod tidy`, can you see anything unusual?

File: go.mod (hunk 2 above)
Comment (jimmidyson) [reply go.mod:19]: Ah yeah good point - thanks for checking that out.

**Comment by jimmidyson** [comment]
@sagikazarmark I was surprised there weren't a lot more removals from `go.sum`. It's not a blocker though.

**Review by jimmidyson — APPROVED** [review]

**Comment by sagikazarmark** [comment]
I'll take a look at it. My guess is there is a transitive dependency on an earlier Kubernetes version.

**Comment by sagikazarmark** [comment]
Controller runtime for instance still relies on 1.18.6

**Comment by jimmidyson** [comment]
@sagikazarmark Ah ok that makes sense

**Review by makkes — APPROVED** [review]

**Review by makkes — APPROVED** [review]

**Comment by jimmidyson** [comment]
@sagikazarmark Sorry for the bother but would you mind rebasing this against master and dropping the last two commits in this branch (only need the updated kubernetes dependencies)? Hopefully that should sort out the CI issues.

**Comment by sagikazarmark** [comment]
@jimmidyson I've just rebased it, looks like at the same time you did. 😄

**Comment by jimmidyson** [comment]
Oh I'm sorry @sagikazarmark, I was just trying to help out to get it in, hope that's ok. I'm going to push another commit just to bump the dependencies to latest now they're published, then we can get this merged. Thank you for kicking this upgrade off.

**Comment by sagikazarmark** [comment]
Sure, no worries. 😄 Let me know if I can help. Thanks!

**Comment by sagikazarmark** [comment]
Looks like it fails with the same error 😕

**Comment by jimmidyson** [comment]
@sagikazarmark My dodgy rebase - I knew I should have waited for you :wink:

**Comment by sagikazarmark** [comment]
Hm, I'm not sure, the rebase looks fine to me. Only `go.mod` and `go.sum` have changes.

**Comment by jimmidyson** [comment]
Yeah it seems to be a consistent test failure.

**Comment by jimmidyson** [comment]
To test the one failing test you can run (after spinning up some clusters and deploying kubefed - `scripts/create-clusters.sh` and `make deploy.kind` should help I think):

```shell
$ make bin/e2e
$ bin/e2e -kubeconfig=${HOME}/.kube/config -ginkgo.v -single-call-timeout=1m -ginkgo.trace -ginkgo.randomizeAllSpecs -ginkgo.focus="deployments.apps.* should be created, read, updated and deleted successfully"
```

**Comment by sagikazarmark** [comment]
I can see a bunch of `ERROR: Cluster object is not as expected` errors caused by an extra `"deployment.kubernetes.io/revision": "1"` annotation in the object.

**Comment by jimmidyson** [comment]
Yeah it looks like a bug in the jsonpatch upgrade (test below).

```go
package a

import (
	"encoding/json"
	"fmt"
	"testing"

	jsonpatch "github.com/evanphx/json-patch"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"

	"sigs.k8s.io/kubefed/pkg/controller/util"
)

func TestPatch(t *testing.T) {
	orig := &unstructured.Unstructured{
		Object: map[string]interface{}{
			"metadata": map[string]interface{}{
				"annotations": map[string]interface{}{
					"deployment.kubernetes.io/revision": "1",
				},
			},
		},
	}
	origJSON, _ := orig.MarshalJSON()

	overrides := util.ClusterOverrides{{
		Op:   "add",
		Path: "/metadata/annotations",
		Value: map[string]string{
			"foo": "bar",
		},
	}, {
		Path: "/metadata/annotations/foo",
		Op:   "remove",
	}}
	jsonPatchBytes, _ := json.Marshal(overrides)
	patch, _ := jsonpatch.DecodePatch(jsonPatchBytes)

	patchedObjectJSONBytes, _ := patch.Apply(origJSON)
	fmt.Println(string(patchedObjectJSONBytes))
}
```

**Comment by jimmidyson** [comment]
Or potentially a fix in the JSONPatch library... I think this is it tbh.

**Comment by jimmidyson** [comment]
Ah yes this is because the override is wrong... fixing now.

**Comment by jimmidyson** [comment]
OK fixed. The override was actually now invalid...

**Comment by jimmidyson** [comment]
Thanks @sagikazarmark!

**Comment by sagikazarmark** [comment]
Thanks for the help @jimmidyson

**Pull request merged**